### DISCUSSION TITLE
Playwright? Pytest? Testing?

### DISCUSSION THREAD

**Discussion by houfu** [reflex-dev/reflex#549] [General]
Has anyone tried to incorporate testing in their pynecone apps using playwright or pytest?

I am curious to find out ideas how to do CI with my pynecone app.

**Comment by Alek99** [comment]
We are currently working on a better way to write integration tests with pynecone. Will be implementing it in the main repo probably by v0.1.18

**Reply by DavidWalshe93** [reply]
Hi @Alek99, I am wondering if there is any update on the testability of Reflex apps? It would be great to have a section in the docs on best practices for testing Reflex apps with Pytest /Playwright on some common use-cases relevant to most sites like component rendering and content checks. 

Would also be great to have some guidance/best practices on how to test the backend API endpoints. 

This would be a big win for Reflex in attracting developers for business/commercial use-cases.

**Reply by venomouse** [reply]
Seconding this. Being able to write test for the framework would be game-changing and set it apart from i.e. Streamlit.

**Reply by picklelo** [reply]
We have a test harness that we expose with `reflex` but we haven't documented it yet. We will add this to the docs soon and write up a tutorial - you can see [for example](https://github.com/reflex-dev/reflex/blob/main/integration/test_upload.py) we have some in our integration tests. They let you send specify events to trigger and then you can do asserts to make sure the UI matches what you expect.

**Reply by ArthurMro** [reply]
@picklelo The link of your example is currently broken. Is there any other example you could provide?

**Reply by picklelo** [reply]
@ArthurMro  See this test / directory: https://github.com/reflex-dev/reflex/blob/main/tests/integration/tests_playwright/test_table.py
